File: web/src/transport/chunkSize.ts
/**
 * Adaptive chunk sizing, protocol.md §5.5:
 *   RTT <20ms, loss <0.1%  -> 1 MB
 *   RTT <100ms, loss <1%   -> 256 KB
 *   worse                  -> 32-64 KB
 * with EWMA-smoothed measurements, discrete steps only (one tier per
 * adjustment), a 5s cooldown between adjustments, and hysteresis
 * (different thresholds for stepping up vs down) to prevent oscillation.
 */

export interface NetworkSample {
  rttMs: number
  lossFraction: number
}

const TIERS = [64 * 1024, 256 * 1024, 1024 * 1024] as const

export class AdaptiveChunkSize {
  private tier = 0
  private smoothedRtt = 0
  private smoothedLoss = 0
  private initialized = false
  private lastAdjustedAt = 0
  private readonly alpha: number
  private readonly cooldownMs: number

  constructor(options: { alpha?: number; cooldownMs?: number } = {}) {
    this.alpha = options.alpha ?? 0.3
    this.cooldownMs = options.cooldownMs ?? 5000
  }

  current(): number {
    // A test hook, in the same shape as the stall windows in
    // transferSession.ts. Climbing a tier honestly takes a 5s cooldown
    // and real getStats() samples, which is exactly the condition that
    // made the second fetch of a session differ from the first — so
    // there has to be some way to reach it without waiting.
    const forced = (globalThis as { DEPOT_FORCE_CHUNK_TIER?: number }).DEPOT_FORCE_CHUNK_TIER
    return forced ?? TIERS[this.tier]
  }

  update(sample: NetworkSample, now: number = Date.now()): void {
    if (!this.initialized) {
      this.smoothedRtt = sample.rttMs
      this.smoothedLoss = sample.lossFraction
      this.initialized = true
      return
    }
    this.smoothedRtt = this.alpha * sample.rttMs + (1 - this.alpha) * this.smoothedRtt
    this.smoothedLoss = this.alpha * sample.lossFraction + (1 - this.alpha) * this.smoothedLoss

    if (now - this.lastAdjustedAt < this.cooldownMs) return

    if (this.tier < TIERS.length - 1 && this.canUpgrade()) {
      this.tier += 1
      this.lastAdjustedAt = now
    } else if (this.tier > 0 && this.shouldDowngrade()) {
      this.tier -= 1
      this.lastAdjustedAt = now
    }
  }

  // Stricter than the spec's nominal thresholds — requires clear headroom
  // before stepping up, so a link hovering right at the boundary doesn't flap.
  private canUpgrade(): boolean {
    if (this.tier === 0) return this.smoothedRtt < 80 && this.smoothedLoss < 0.008
    if (this.tier === 1) return this.smoothedRtt < 15 && this.smoothedLoss < 0.0008
    return false
  }

  // At or looser than the spec's nominal thresholds — reacts to degradation promptly.
  private shouldDowngrade(): boolean {
    if (this.tier === 2) return this.smoothedRtt >= 20 || this.smoothedLoss >= 0.001
    if (this.tier === 1) return this.smoothedRtt >= 100 || this.smoothedLoss >= 0.01
    return false
  }
}

/**
 * Reads RTT off the active candidate pair. Standard getStats() doesn't
 * expose a reliable cross-browser loss fraction for SCTP data channels, so
 * loss is reported as 0 (optimistic) rather than overclaiming precision —
 * sizing here leans on RTT, with loss folded in wherever a browser does
 * report it.
 */
export async function sampleNetwork(pc: RTCPeerConnection): Promise<NetworkSample | null> {
  const report = await pc.getStats()
  for (const stat of report.values()) {
    const pair = stat as RTCIceCandidatePairStats
    if (pair.type === 'candidate-pair' && pair.state === 'succeeded' && (pair.nominated ?? true)) {
      const rttMs = typeof pair.currentRoundTripTime === 'number' ? pair.currentRoundTripTime * 1000 : 0
      return { rttMs, lossFraction: 0 }
    }
  }
  return null
}

/** CDC parameters scaled to a target average chunk size, keeping min/max proportionally bounded. */
/**
 * CDC parameters for a target average, never exceeding what CAPS agreed.
 *
 * [ceiling] matters more than it looks. A chunker asked for an average
 * of N produces chunks of up to 4N — that spread is how content-defined
 * chunking finds its boundaries — while §5.4 makes the negotiated
 * maxChunkSize a hard limit that *both* sides reject a manifest for
 * exceeding. Deriving the parameters from the average alone therefore
 * built manifests that the peer was entitled to refuse, and did:
 *
 *  - Uploading passed the negotiated size straight in as the average,
 *    so every chunk four times over the limit. Sending anything from
 *    the browser failed with "chunk 0 is larger than CAPS agreed".
 *  - Serving passed §5.5's current tier. That starts at 64 KB, whose
 *    maximum of 256 KB is comfortably inside a 1 MB ceiling — so the
 *    first fetch of a session worked. Once the link measured well and
 *    the tier climbed to 1 MB, the maximum became 4 MB and every fetch
 *    after that was refused, until the page was reloaded and the tier
 *    reset. "The first one works and then I have to reload" was this.
 *
 * So the ceiling caps the maximum, and the average follows it down
 * rather than the other way about.
 */
export function cdcParamsForAvg(
  avgSize: number,
  ceiling = Number.POSITIVE_INFINITY,
): { minSize: number; avgSize: number; maxSize: number } {
  const maxSize = Math.min(avgSize * 4, ceiling)
  // Half the maximum rather than a quarter of it when the ceiling binds.
  // The chunker's own spread is four times the average, so a quarter
  // would be the figure that never truncates — but truncating the tail
  // of the distribution costs almost nothing here (chunks are used for
  // resuming one transfer, not for dedup across versions) and a quarter
  // would double the number of messages for no gain anyone can measure.
  const avg = Math.max(1, Math.min(avgSize, Math.floor(maxSize / 2)))
  return { minSize: Math.max(1, Math.floor(avg / 4)), avgSize: avg, maxSize }
}

File: android/app/src/main/java/com/depot/app/transport/ChunkSize.kt
package com.depot.app.transport

import kotlin.coroutines.resume
import kotlinx.coroutines.suspendCancellableCoroutine
import org.webrtc.PeerConnection
import org.webrtc.RTCStatsCollectorCallback
import org.webrtc.RTCStatsReport

/**
 * Adaptive chunk sizing, protocol.md §5.5:
 *   RTT <20ms, loss <0.1%  -> 1 MB
 *   RTT <100ms, loss <1%   -> 256 KB
 *   worse                  -> 32-64 KB
 *
 * EWMA-smoothed measurements, one discrete step per adjustment, a 5s
 * cooldown, and different thresholds for stepping up than down. All four
 * rules exist for the same reason: a link sitting right on a boundary
 * would otherwise flap between tiers, and re-chunking a file is not free.
 *
 * Deliberately the same constants as the web implementation's
 * chunkSize.ts. A Depot and a Client that disagreed about sizing would
 * still interoperate — CAPS bounds it and the manifest carries the actual
 * lengths — but they would behave differently on the same link for no
 * reason anyone could see.
 */
data class NetworkSample(val rttMs: Double, val lossFraction: Double)

private val TIERS = intArrayOf(64 * 1024, 256 * 1024, 1024 * 1024)

class AdaptiveChunkSize(
    private val alpha: Double = 0.3,
    private val cooldownMs: Long = 5_000,
) {
    private var tier = 0
    private var smoothedRtt = 0.0
    private var smoothedLoss = 0.0
    private var initialized = false
    private var lastAdjustedAt = 0L

    fun current(): Int = TIERS[tier]

    @Synchronized
    fun update(sample: NetworkSample, now: Long = System.currentTimeMillis()) {
        if (!initialized) {
            smoothedRtt = sample.rttMs
            smoothedLoss = sample.lossFraction
            initialized = true
            return
        }
        smoothedRtt = alpha * sample.rttMs + (1 - alpha) * smoothedRtt
        smoothedLoss = alpha * sample.lossFraction + (1 - alpha) * smoothedLoss

        if (now - lastAdjustedAt < cooldownMs) return

        if (tier < TIERS.size - 1 && canUpgrade()) {
            tier += 1
            lastAdjustedAt = now
        } else if (tier > 0 && shouldDowngrade()) {
            tier -= 1
            lastAdjustedAt = now
        }
    }

    /** Stricter than the nominal thresholds: clear headroom before stepping up. */
    private fun canUpgrade(): Boolean = when (tier) {
        0 -> smoothedRtt < 80 && smoothedLoss < 0.008
        1 -> smoothedRtt < 15 && smoothedLoss < 0.0008
        else -> false
    }

    /** At or looser than nominal: react to degradation promptly. */
    private fun shouldDowngrade(): Boolean = when (tier) {
        2 -> smoothedRtt >= 20 || smoothedLoss >= 0.001
        1 -> smoothedRtt >= 100 || smoothedLoss >= 0.01
        else -> false
    }
}

/**
 * RTT off the active candidate pair.
 *
 * getStats() exposes no dependable loss fraction for an SCTP data channel,
 * so loss is reported as zero rather than inventing a number sizing cannot
 * actually justify — the same call the web side makes, and the same
 * admission.
 */
suspend fun sampleNetwork(pc: PeerConnection): NetworkSample? {
    val report: RTCStatsReport = suspendCancellableCoroutine { cont ->
        pc.getStats(RTCStatsCollectorCallback { cont.resume(it) })
    }
    val pair = report.statsMap.values.firstOrNull { s ->
        s.type == "candidate-pair" && s.members["state"] == "succeeded"
    } ?: return null
    val rtt = when (val value = pair.members["currentRoundTripTime"]) {
        is Number -> value.toDouble() * 1000.0
        else -> 0.0
    }
    return NetworkSample(rttMs = rtt, lossFraction = 0.0)
}
